helix-term/src/ui/picker.rs

```diff
@@ -7,17 +10,150 @@ use tui::{
 
 use fuzzy_matcher::skim::SkimMatcherV2 as Matcher;
 use fuzzy_matcher::FuzzyMatcher;
+use tui::widgets::Widget;
 
-use std::borrow::Cow;
+use std::{borrow::Cow, collections::HashMap, path::PathBuf};
 
 use crate::ui::{Prompt, PromptEvent};
-use helix_core::Position;
+use helix_core::{Position, Selection};
 use helix_view::{
+    document::canonicalize_path,
     editor::Action,
     graphics::{Color, CursorKind, Rect, Style},
-    Editor,
+    Document, Editor, View, ViewId,
 };
 
+/// File path and line number
+type FileLocation = (PathBuf, usize);
+
+pub struct FilePicker<T> {
+    picker: Picker<T>,
+    /// Caches paths to documents
+    preview_cache: HashMap<PathBuf, Document>,
+    /// Given an item in the picker, return the file path and line number to display.
+    file_fn: Box<dyn Fn(&Editor, &T) -> Option<FileLocation>>,
+    // A view id to be shared by all documents in the cache. Mostly a hack since a doc
+    // requires at least one selection.
+    _preview_view_id: ViewId,
+}
+
+impl<T> FilePicker<T> {
+    pub fn new(
+        options: Vec<T>,
+        format_fn: impl Fn(&T) -> Cow<str> + 'static,
+        callback_fn: impl Fn(&mut Editor, &T, Action) + 'static,
+        preview_fn: impl Fn(&Editor, &T) -> Option<FileLocation> + 'static,
+    ) -> Self {
+        Self {
+            picker: Picker::new(options, format_fn, callback_fn),
+            preview_cache: HashMap::new(),
+            file_fn: Box::new(preview_fn),
+            _preview_view_id: ViewId::default(),
+        }
+    }
+
+    fn current_file(&self, editor: &Editor) -> Option<FileLocation> {
+        self.picker
+            .selection()
+            .and_then(|current| (self.file_fn)(editor, current))
+            .and_then(|(path, line)| canonicalize_path(&path).ok().zip(Some(line)))
+    }
+
+    fn calculate_preview(&mut self, editor: &Editor) {
+        if let Some((path, _line)) = self.current_file(editor) {
+            if !self.preview_cache.contains_key(&path) && editor.document_by_path(&path).is_none() {
+                let mut doc =
+                    Document::open(&path, None, Some(&editor.theme), Some(&editor.syn_loader))
+                        .unwrap();
+                // HACK: a doc needs atleast one selection
+                doc.set_selection(self._preview_view_id, Selection::point(0));
+                self.preview_cache.insert(path, doc);
+            }
+        }
+    }
+}
```

Comment:
This is the part that should be made async; how should I go about doing so ? I haven't written that much async code 😅

Reply:
This is the hard part, or maybe for now you just don't show highlighting.

Reply:
I disabled highlighting for now since the treesitter parsing will also need to be done async (files already loaded in memory will be highlighted though).